### PR TITLE
adding check to prevent tool cross talk

### PR DESCRIPTION
**Pull request by wjwwood**
See: https://github.com/ros/catkin/issues/742

### CODE REVIEW

**Comment by tfoote** [comment]
+1

The unit tests will need to be rerun once catkin_pkg is released with the new version.

**Comment by jbohren** [comment]
Should this be done here or in `build_isolated_workspace()` in `build.py`?

**Comment by wjwwood** [comment]
I wanted to have the check occur before the `dry_run` function, now I realize that the `mark_space_as_built_by` functions should also be after the `dry_run` function.

**Comment by jbohren** [comment]
Ok then we should probably move the other check here too.

**Comment by dirk-thomas** [comment]
This has been addressed in `catkin_pkg` and `catkin` which have both been released since then. I have closed the related tickets in these repos. Once `catkin_tools` merges and releases this check it would actually work.

File: catkin_tools/verbs/catkin_build/cli.py [hunk 1]
```diff
@@ -142,6 +143,8 @@ def prepare_arguments(parser):
         help='Adds a build summary to the end of a build; defaults to on with --continue-on-failure, off otherwise')
     add('--no-summarize', '--no-summary', action='store_false', dest='summarize',
         help='explicitly disable the end of build summary')
+    add('--override-build-tool-check',
```
Comment (NikolausDemmel) [inline catkin_tools/verbs/catkin_build/cli.py]: shouldn't there be `, action='store_true', default=False,`?

**Comment by NikolausDemmel** [comment]
> Ok then we should probably move the other check here too.

which check exactly?

> I wanted to have the check occur before the dry_run function, now I realize that the mark_space_as_built_by functions should also be after the dry_run function.

Why does it need to be moved after `dry_run`?

**Comment by NikolausDemmel** [comment]
Could one of the owners retrigger the tests, please?

File: catkin_tools/verbs/catkin_build/cli.py (hunk 1 above)
Comment (wjwwood) [inline catkin_tools/verbs/catkin_build/cli.py]: Yes, 7b7ffc0.

**Comment by wjwwood** [comment]
> Why does it need to be moved after dry_run?

Because otherwise running `catkin build -n` on an empty workspace will result in the creation of a build and devel folder marked as catkin build's but with nothing else in them.

So now this is the case:

- Run catkin_make, succeeds
- Run catkin build, fails
- Run catkin build -n, fails
- Run catkin build -n --override-build-tool-check, warns but doesn't build
- Run catkin build -n --override-build-tool-check, _still_ warns but doesn't build
- Run catkin build --override-build-tool-check, warns but builds
- Run catkin build --override-build-tool-check, no warning and builds
- rm -rf build devel
- Run catkin build -n, no build or devel folder created
- Run catkin build , builds and therefore build and devel folder created

I've tested this with both `catkin_make` and `catkin_make_isolated` locally. Works for me. I also removed the extra checks which I think was what @jbohren was talking about, but he can comment if that wasn't the case.

**Pull request merged**